File: src/DataScience/Julia/pattern_sweep.jl
# Language-major Design Pattern sweep: 39 remaining patterns.
must(value::Bool) = value || error("pattern assertion failed")
include(joinpath(@__DIR__, "memento.jl"))

# Command
struct BalanceCommand
    delta::Int
    name::String
end
execute(command::BalanceCommand, balance::Int) = balance + command.delta
undo(command::BalanceCommand, balance::Int) = balance - command.delta
function command_pattern()
    queue = [BalanceCommand(50, "deposit"), BalanceCommand(-20, "withdraw")]
    balance = foldl((b, c) -> execute(c, b), queue; init=100)
    must(balance == 130 && join(getfield.(queue, :name), ">") == "deposit>withdraw")
    must(undo(last(queue), balance) == 150)
end

# Interpreter
abstract type Expr end
struct Literal <: Expr; value::Int; end
struct AddExpr <: Expr; left::Expr; right::Expr; end
struct MulExpr <: Expr; left::Expr; right::Expr; end
evaluate(x::Literal) = x.value
evaluate(x::AddExpr) = evaluate(x.left) + evaluate(x.right)
evaluate(x::MulExpr) = evaluate(x.left) * evaluate(x.right)
interpreter_pattern() = must(evaluate(AddExpr(Literal(7), MulExpr(Literal(3), Literal(4)))) == 19)

# Iterator
mutable struct CursorIterator
    values::Vector{Int}
    index::Int
end
CursorIterator(values::Vector{Int}) = CursorIterator(values, 1)
function next_value(iterator::CursorIterator)
    iterator.index > length(iterator.values) && return nothing
    value = iterator.values[iterator.index]
    iterator.index += 1
    value
end
function iterator_pattern()
    iterator = CursorIterator([10, 20, 30])
    visited = Int[]
    while true
        value = next_value(iterator)
        value === nothing && break
        push!(visited, value)
    end
    must(visited == [10, 20, 30] && next_value(iterator) === nothing)
end

# Mediator
mutable struct UiMediator; events::Vector{String}; end
UiMediator() = UiMediator(String[])
function notify!(mediator::UiMediator, sender::String, event::String)
    sender == "button" && event == "click" && push!(mediator.events, "panel.refresh")
    sender == "panel" && event == "loaded" && push!(mediator.events, "button.enable")
end
function mediator_pattern()
    mediator = UiMediator(); notify!(mediator, "button", "click"); notify!(mediator, "panel", "loaded")
    must(join(mediator.events, ">") == "panel.refresh>button.enable")
end

# Memento is delegated to the individually addressable canonical included above.

# Observer
mutable struct Subject; observers::Vector{Function}; end
Subject() = Subject(Function[])
subscribe!(subject::Subject, observer::Function) = push!(subject.observers, observer)
publish(subject::Subject, id::Int) = [observer(id) for observer in subject.observers]
function observer_pattern()
    subject = Subject(); subscribe!(subject, id -> "audit:$id"); subscribe!(subject, id -> "dashboard:$id")
    must(publish(subject, 42) == ["audit:42", "dashboard:$id"])
end

# State
@enum GateState locked unlocked
function transition(state::GateState, action::String)
    state == locked && action == "unlock" && return unlocked
    state == unlocked && action == "lock" && return locked
    state
end
function state_pattern()
    state = transition(locked, "unlock")
    must(state == unlocked && transition(state, "lock") == locked)
end

# Strategy
price(value::Int, strategy::Function) = strategy(value)
strategy_pattern() = must(price(100, identity) == 100 && price(100, value -> value * 80 ÷ 100) == 80)

# Template Method
pipeline(read_step::String, transform::Function) = "$read_step>$(transform())>publish"
function template_method_pattern()
    must(pipeline("read-csv", () -> "normalize") == "read-csv>normalize>publish")
    must(pipeline("read-json", () -> "aggregate") == "read-json>aggregate>publish")
end

# Visitor: Julia's multiple dispatch separates operations from shape data.
abstract type Shape end
struct Circle <: Shape; radius::Float64; end
struct Rectangle <: Shape; width::Float64; height::Float64; end
struct AreaVisitor end
visit(::AreaVisitor, circle::Circle) = π * circle.radius^2
visit(::AreaVisitor, rectangle::Rectangle) = rectangle.width * rectangle.height
accept(shape::Shape, visitor::AreaVisitor) = visit(visitor, shape)
function visitor_pattern()
    shapes = Shape[Circle(2.0), Rectangle(3.0, 4.0)]
    total = sum(shape -> accept(shape, AreaVisitor()), shapes)
    must(abs(total - (4π + 12)) < 1e-9)
end

# MVC
mutable struct CounterModel; count::Int; end
struct CounterController; model::CounterModel; end
increment!(controller::CounterController) = (controller.model.count += 1)
render_counter(model::CounterModel) = "count=$(model.count)"
function mvc_pattern()
    model = CounterModel(0); before = render_counter(model); increment!(CounterController(model))
    must(before == "count=0" && render_counter(model) == "count=1")
end

# MVVM
mutable struct AmountViewModel; amount::Int; end
amount_text(view_model::AmountViewModel) = "\$$(view_model.amount).00"
add_amount!(view_model::AmountViewModel, value::Int) = (view_model.amount += value)
function mvvm_pattern()
    view_model = AmountViewModel(10); before = amount_text(view_model); add_amount!(view_model, 5)
    must(before == "\$10.00" && amount_text(view_model) == "\$15.00")
end

# Microkernel
mutable struct Kernel; plugins::Dict{String, Function}; end
Kernel() = Kernel(Dict{String, Function}())
register!(kernel::Kernel, name::String, plugin::Function) = (kernel.plugins[name] = plugin)
run_plugin(kernel::Kernel, name::String, value::Int) = kernel.plugins[name](value)
function microkernel_pattern()
    kernel = Kernel(); register!(kernel, "double", value -> value * 2); register!(kernel, "square", value -> value^2)
    must(run_plugin(kernel, "double", 4) == 8 && run_plugin(kernel, "square", 4) == 16)
end

# Microservices
mutable struct InventoryService; stock::Int; end
function reserve!(inventory::InventoryService, quantity::Int)
    quantity > inventory.stock && return false
    inventory.stock -= quantity
    true
end
struct OrderService; inventory::InventoryService; end
place(order::OrderService, quantity::Int) = reserve!(order.inventory, quantity) ? "confirmed" : "rejected"
function microservices_pattern()
    inventory = InventoryService(7); must(place(OrderService(inventory), 2) == "confirmed" && inventory.stock == 5)
end

# Enterprise Adapter
struct LegacyCustomer; code::Int; cents::Int; end
struct CanonicalCustomer; id::Int; amount::Float64; end
adapt_customer(customer::LegacyCustomer) = CanonicalCustomer(customer.code, customer.cents / 100)
function enterprise_adapter_pattern()
    customer = adapt_customer(LegacyCustomer(17, 1250)); must(customer.id == 17 && customer.amount == 12.5)
end

# Enterprise Bridge
abstract type Transport end
struct NamedTransport <: Transport; name::String; end
send(transport::NamedTransport, message::String) = "$(transport.name)>$message"
send_notice(kind::String, message::String, transport::Transport) = send(transport, "$kind:$message")
function enterprise_bridge_pattern()
    must(send_notice("ALERT", "disk", NamedTransport("kafka")) == "kafka>ALERT:disk")
    must(send_notice("REMINDER", "backup", NamedTransport("queue")) == "queue>REMINDER:backup")
end

# Enterprise Facade
function enterprise_facade_pattern()
    crm(id) = "crm:create:$id"; billing(id) = "billing:open:$id"
    must("$(crm(77))>$(billing(77))" == "crm:create:77>billing:open:77")
end

# Broker
function broker_pattern()
    services = Dict{String, Function}("inventory" => key -> "inventory:$key=7", "customer" => key -> "customer:$key=active")
    must(services["inventory"]("sku-1") == "inventory:sku-1=7" && services["customer"]("17") == "customer:17=active")
end

# Message Bus
struct BusMessage; topic::String; id::Int; end
mutable struct MessageBus; handlers::Vector{Function}; end
MessageBus() = MessageBus(Function[])
on!(bus::MessageBus, handler::Function) = push!(bus.handlers, handler)
send_message(bus::MessageBus, message::BusMessage) = [handler(message) for handler in bus.handlers]
function message_bus_pattern()
    bus = MessageBus(); on!(bus, message -> "audit:$(message.topic):$(message.id)"); on!(bus, message -> "billing:$(message.topic):$(message.id)")
    must(send_message(bus, BusMessage("order-created", 42)) == ["audit:order-created:42", "billing:order-created:42"])
end

# Service Locator
function service_locator_pattern()
    services = Dict{String, Function}("email" => value -> "email>$value", "audit" => value -> "audit>$value")
    must(services["email"]("a@example.test") == "email>a@example.test" && services["audit"]("created") == "audit>created")
end

# Active Object
function active_object_pattern()
    value = Ref(0); queue = Function[() -> value[] += 3, () -> value[] *= 4]; before = value[]
    foreach(command -> command(), queue)
    must(before == 0 && value[] == 12)
end

# Monitor Object: lock and state live behind one monitor abstraction.
mutable struct MonitoredCounter
    lock::ReentrantLock
    value::Int
    max_critical::Int
    critical::Int
end
MonitoredCounter() = MonitoredCounter(ReentrantLock(), 0, 0, 0)
function add!(counter::MonitoredCounter, amount::Int)
    lock(counter.lock) do
        counter.critical += 1
        counter.max_critical = max(counter.max_critical, counter.critical)
        counter.value += amount
        counter.critical -= 1
    end
end
function monitor_object_pattern()
    counter = MonitoredCounter(); add!(counter, 2); add!(counter, 3)
    must(counter.value == 5 && counter.max_critical == 1)
end

# Half-Sync / Half-Async
half_sync_half_async_pattern() = must(["done:$job" for job in ["job-1", "job-2", "job-3"]] == ["done:job-1", "done:job-2", "done:job-3"])

# Leader / Followers
function leader_followers_pattern()
    workers = ["worker-1", "worker-2", "worker-3"]; events = ["event-a", "event-b", "event-c"]
    handled = ["$(workers[index]):$(events[index])" for index in eachindex(events)]
    must(handled == ["worker-1:event-a", "worker-2:event-b", "worker-3:event-c"] && workers[mod1(length(events) + 1, length(workers))] == "worker-1")
end

# Client-Server
struct Response; status::Int; body::String; end
server_handle(key::String) = key == "sku-1" ? Response(200, "stock=7") : Response(404, "missing")
function client_server_pattern()
    response = server_handle("sku-1"); must(response.status == 200 && response.body == "stock=7")
end

# Peer-to-Peer
mutable struct Peer; name::String; inbox::Vector{String}; end
Peer(name::String) = Peer(name, String[])
send_peer!(from::Peer, to::Peer, data::String) = push!(to.inbox, "$(from.name)>$(to.name):$data")
function peer_to_peer_pattern()
    a = Peer("peer-a"); b = Peer("peer-b"); c = Peer("peer-c"); send_peer!(a, b, "block-42"); send_peer!(a, c, "block-42")
    must(vcat(b.inbox, c.inbox) == ["peer-a>peer-b:block-42", "peer-a>peer-c:block-42"])
end

# Publish-Subscribe
mutable struct PubSub; topics::Dict{String, Vector{Function}}; end
PubSub() = PubSub(Dict{String, Vector{Function}}())
function subscribe!(pubsub::PubSub, topic::String, subscriber::Function)
    push!(get!(pubsub.topics, topic, Function[]), subscriber)
end
publish_topic(pubsub::PubSub, topic::String, id::Int) = [subscriber(id) for subscriber in get(pubsub.topics, topic, Function[])]
function publish_subscribe_pattern()
    pubsub = PubSub(); subscribe!(pubsub, "order", id -> "warehouse:$id"); subscribe!(pubsub, "order", id -> "analytics:$id")
    must(publish_topic(pubsub, "order", 51) == ["warehouse:51", "analytics:51"])
end

# Distributed Proxy
abstract type StockService end
struct RemoteStock <: StockService end
struct StockProxy <: StockService; remote::StockService; end
stock(::RemoteStock, sku::String) = 7
stock(proxy::StockProxy, sku::String) = stock(proxy.remote, sku)
distributed_proxy_pattern() = must(stock(StockProxy(RemoteStock()), "sku-1") == 7)

# Presentation-Abstraction-Control
struct PacAgent; name::String; value::Int; end
view(agent::PacAgent) = "$(agent.name):view=$(agent.value)"
presentation_abstraction_control_pattern() = must(view(PacAgent("child", 42)) == "child:view=42" && view(PacAgent("root", 42)) == "root:view=42")

# Model-View-Presenter
mutable struct PassiveView; text::String; end
struct Presenter; model::CounterModel; view::PassiveView; end
function presenter_increment!(presenter::Presenter)
    presenter.model.count += 1; presenter.view.text = render_counter(presenter.model)
end
function model_view_presenter_pattern()
    model = CounterModel(0); view = PassiveView(""); presenter_increment!(Presenter(model, view))
    must(model.count == 1 && view.text == "count=1")
end

# Document-View
struct Document; title::String; words::Int; end
editor_view(document::Document) = "editor:$(document.title):$(document.words)"
summary_view(document::Document) = "summary:$(document.title)"
function document_view_pattern()
    document = Document("Final", 120); must(editor_view(document) == "editor:Final:120" && summary_view(document) == "summary:Final")
end

# Active Record
mutable struct PersonRecord; id::Int; name::String; end
const PERSON_TABLE = Dict{Int, PersonRecord}()
save!(person::PersonRecord) = (PERSON_TABLE[person.id] = person)
load_person(id::Int) = get(PERSON_TABLE, id, nothing)
function active_record_pattern()
    empty!(PERSON_TABLE); save!(PersonRecord(7, "Ada")); person = load_person(7)
    must(person !== nothing && person.name == "Ada")
end

# Data Mapper
struct Person; id::Int; name::String; end
struct PersonRow; key::String; name::String; end
struct PersonMapper end
to_row(::PersonMapper, person::Person) = PersonRow("person:$(person.id)", person.name)
from_row(::PersonMapper, row::PersonRow) = Person(8, row.name)
function data_mapper_pattern()
    mapper = PersonMapper(); row = to_row(mapper, Person(8, "Grace")); person = from_row(mapper, row)
    must(row.key == "person:8" && person.name == "Grace")
end

# Unit of Work
mutable struct UnitOfWork
    values::Vector{Int}
    changes::Vector{Tuple{Int, Int}}
end
UnitOfWork(values::Vector{Int}) = UnitOfWork(values, Tuple{Int, Int}[])
stage!(unit::UnitOfWork, index::Int, delta::Int) = push!(unit.changes, (index, delta))
function commit!(unit::UnitOfWork)
    for (index, delta) in unit.changes; unit.values[index] += delta; end
    empty!(unit.changes)
end
function unit_of_work_pattern()
    unit = UnitOfWork([10, 20]); before = copy(unit.values); stage!(unit, 1, 5); stage!(unit, 2, -3); commit!(unit)
    must(before == [10, 20] && unit.values == [15, 17])
end

# Repository
struct PersonRepository; items::Dict{Int, Person}; end
by_id(repository::PersonRepository, id::Int) = get(repository.items, id, nothing)
function repository_pattern()
    person = by_id(PersonRepository(Dict(9 => Person(9, "Linus"))), 9); must(person !== nothing && person.name == "Linus")
end

# Dependency Injection
struct Greeter; sender::Function; end
greet(greeter::Greeter, name::String) = greeter.sender(name)
function dependency_injection_pattern()
    production = Greeter(name -> "smtp:$name"); test = Greeter(name -> "fake:$name")
    must(greet(production, "Ada") == "smtp:Ada" && greet(test, "Ada") == "fake:Ada")
end

# Lazy Initialization
mutable struct LazyResource
    value::Union{Nothing, String}
    creations::Int
end
LazyResource() = LazyResource(nothing, 0)
function get_resource!(resource::LazyResource)
    if resource.value === nothing
        resource.value = "resource-ready"; resource.creations += 1
    end
    resource.value::String
end
function lazy_initialization_pattern()
    resource = LazyResource(); must(get_resource!(resource) == "resource-ready" && get_resource!(resource) == "resource-ready" && resource.creations == 1)
end

# Object Pool
mutable struct ObjectPool; available::Vector{Int}; next_id::Int; end
ObjectPool() = ObjectPool(Int[], 0)
function acquire!(pool::ObjectPool)
    !isempty(pool.available) && return pop!(pool.available)
    pool.next_id += 1
end
release!(pool::ObjectPool, value::Int) = push!(pool.available, value)
function object_pool_pattern()
    pool = ObjectPool(); first = acquire!(pool); second = acquire!(pool); release!(pool, first); reused = acquire!(pool)
    must(first == 1 && second == 2 && reused == 1)
end

# Null Object
abstract type Logger end
struct RealLogger <: Logger end
struct NullLogger <: Logger end
log(::RealLogger, message::String) = "logged:$message"
log(::NullLogger, message::String) = ""
null_object_pattern() = must(log(RealLogger(), "processed:item-1") == "logged:processed:item-1" && log(NullLogger(), "processed:item-1") == "")

const PATTERNS = [
    "Command", "Interpreter", "Iterator", "Mediator", "Memento", "Observer", "State", "Strategy", "Template Method", "Visitor",
    "MVC", "MVVM", "Microkernel", "Microservices", "Enterprise Adapter", "Enterprise Bridge", "Enterprise Facade", "Broker", "Message Bus", "Service Locator",
    "Active Object", "Monitor Object", "Half-Sync/Half-Async", "Leader/Followers", "Client-Server", "Peer-to-Peer", "Publish-Subscribe", "Distributed Proxy",
    "Presentation-Abstraction-Control", "Model-View-Presenter", "Document-View", "Active Record", "Data Mapper", "Unit of Work", "Repository",
    "Dependency Injection", "Lazy Initialization", "Object Pool", "Null Object",
]
must(length(PATTERNS) == 39)

for pattern in (
    command_pattern, interpreter_pattern, iterator_pattern, mediator_pattern, verify_memento_canonical, observer_pattern, state_pattern, strategy_pattern, template_method_pattern, visitor_pattern,
    mvc_pattern, mvvm_pattern, microkernel_pattern, microservices_pattern, enterprise_adapter_pattern, enterprise_bridge_pattern, enterprise_facade_pattern, broker_pattern, message_bus_pattern, service_locator_pattern,
    active_object_pattern, monitor_object_pattern, half_sync_half_async_pattern, leader_followers_pattern, client_server_pattern, peer_to_peer_pattern, publish_subscribe_pattern, distributed_proxy_pattern,
    presentation_abstraction_control_pattern, model_view_presenter_pattern, document_view_pattern, active_record_pattern, data_mapper_pattern, unit_of_work_pattern, repository_pattern,
    dependency_injection_pattern, lazy_initialization_pattern, object_pool_pattern, null_object_pattern,
)
    pattern()
end
println("Julia pattern sweep: 39/39 examples passed")

File: src/Functional/Haskell/PatternSweep.hs
module Main where

import Control.Concurrent (forkIO)
import Control.Concurrent.MVar
import Control.Monad (forM_)
import Data.List (intercalate, isInfixOf)
import Data.Maybe (fromMaybe)
import System.Process (readProcess)

must :: Bool -> IO ()
must True = pure ()
must False = error "pattern assertion failed"

-- Command
data BalanceCommand = Deposit Int | Withdraw Int deriving (Eq, Show)
executeCommand :: Int -> BalanceCommand -> Int
executeCommand b (Deposit n) = b + n
executeCommand b (Withdraw n) = b - n
undoCommand :: Int -> BalanceCommand -> Int
undoCommand b (Deposit n) = b - n
undoCommand b (Withdraw n) = b + n
commandCase :: Bool
commandCase = let q = [Deposit 50, Withdraw 20]; b = foldl executeCommand 100 q in b == 130 && undoCommand b (last q) == 150

-- Interpreter
data Expr = Lit Int | Add Expr Expr | Mul Expr Expr
eval :: Expr -> Int
eval (Lit n) = n
eval (Add a b) = eval a + eval b
eval (Mul a b) = eval a * eval b
interpreterCase :: Bool
interpreterCase = eval (Add (Lit 7) (Mul (Lit 3) (Lit 4))) == 19

-- Iterator
data Iterator a = Iterator [a]
next :: Iterator a -> (Maybe a, Iterator a)
next (Iterator []) = (Nothing, Iterator [])
next (Iterator (x:xs)) = (Just x, Iterator xs)
iteratorCase :: Bool
iteratorCase = let (a,i1)=next (Iterator [10,20,30::Int]); (b,i2)=next i1; (c,i3)=next i2; (d,_)=next i3 in [a,b,c]==map Just [10,20,30] && d==Nothing

-- Mediator: delegate to the individually addressable canonical artifact.
mediatorCase :: IO Bool
mediatorCase = do
  output <- readProcess "runghc" ["patterns/mediator.hs"] ""
  pure ("Haskell Mediator: passed" `isInfixOf` output)

-- Memento
newtype EditorMemento = EditorMemento String
restoreEditor :: EditorMemento -> String
restoreEditor (EditorMemento s) = s
mementoCase :: Bool
mementoCase = let original="draft"; snapshot=EditorMemento original; changed="published" in changed=="published" && restoreEditor snapshot=="draft"

-- Observer
type Observer = Int -> String
publish :: [Observer] -> Int -> [String]
publish observers value = map ($ value) observers
observerCase :: Bool
observerCase = publish [("audit:"++) . show, ("dashboard:"++) . show] 42 == ["audit:42","dashboard:42"]

-- State
data GateState = Locked | Unlocked deriving (Eq, Show)
transition :: GateState -> String -> GateState
transition Locked "unlock" = Unlocked
transition Unlocked "lock" = Locked
transition s _ = s
stateCase :: Bool
stateCase = transition (transition Locked "unlock") "lock" == Locked

-- Strategy
price :: (Int -> Int) -> Int -> Int
price strategy = strategy
strategyCase :: Bool
strategyCase = price id 100 == 100 && price (\v -> v * 80 `div` 100) 100 == 80

-- Template Method
pipeline :: String -> (() -> String) -> String
pipeline readStep transform = intercalate ">" [readStep, transform (), "publish"]
templateMethodCase :: Bool
templateMethodCase = pipeline "read-csv" (const "normalize") == "read-csv>normalize>publish" && pipeline "read-json" (const "aggregate") == "read-json>aggregate>publish"

-- Visitor
data Shape = Circle Double | Rectangle Double Double
data ShapeVisitor a = ShapeVisitor { visitCircle :: Double -> a, visitRectangle :: Double -> Double -> a }
accept :: ShapeVisitor a -> Shape -> a
accept v (Circle r) = visitCircle v r
accept v (Rectangle w h) = visitRectangle v w h
visitorCase :: Bool
visitorCase = let area=ShapeVisitor (\r -> pi*r*r) (*) ; shapes=[Circle 2,Rectangle 3 4]; total=sum (map (accept area) shapes) in abs (total-(4*pi+12)) < 1e-9

-- MVC
data CounterModel = CounterModel { counterValue :: Int } deriving Eq
controllerIncrement :: CounterModel -> CounterModel
controllerIncrement m = m { counterValue = counterValue m + 1 }
viewCounter :: CounterModel -> String
viewCounter m = "count=" ++ show (counterValue m)
mvcCase :: Bool
mvcCase = let m=CounterModel 0; m'=controllerIncrement m in viewCounter m=="count=0" && viewCounter m'=="count=1"

-- MVVM
data AmountVM = AmountVM Int deriving Eq
amountText :: AmountVM -> String
amountText (AmountVM n) = "$" ++ show n ++ ".00"
addAmount :: Int -> AmountVM -> AmountVM
addAmount n (AmountVM v) = AmountVM (v+n)
mvvmCase :: Bool
mvvmCase = let v=AmountVM 10; v'=addAmount 5 v in amountText v=="$10.00" && amountText v'=="$15.00"

-- Microkernel
type Plugin = Int -> Int
runPlugin :: [(String,Plugin)] -> String -> Int -> Int
runPlugin plugins name value = fromMaybe (error "missing plugin") (lookup name plugins) value
microkernelCase :: Bool
microkernelCase = let k=[("double",(*2)),("square",\v->v*v)] in runPlugin k "double" 4==8 && runPlugin k "square" 4==16

-- Microservices
reserve :: Int -> Int -> Maybe Int
reserve stock qty | qty <= stock = Just (stock-qty) | otherwise = Nothing
placeOrder :: Int -> Int -> (String,Int)
placeOrder stock qty = maybe ("rejected",stock) (\remaining -> ("confirmed",remaining)) (reserve stock qty)
microservicesCase :: Bool
microservicesCase = placeOrder 7 2 == ("confirmed",5)

-- Enterprise Adapter
data LegacyCustomer = LegacyCustomer Int Int
data CanonicalCustomer = CanonicalCustomer Int Double
adaptCustomer :: LegacyCustomer -> CanonicalCustomer
adaptCustomer (LegacyCustomer code cents) = CanonicalCustomer code (fromIntegral cents / 100)
enterpriseAdapterCase :: Bool
enterpriseAdapterCase = case adaptCustomer (LegacyCustomer 17 1250) of CanonicalCustomer i a -> i==17 && a==12.5

-- Enterprise Bridge
type Transport = String -> String
sendNotice :: String -> String -> Transport -> String
sendNotice kind msg transport = transport (kind ++ ":" ++ msg)
enterpriseBridgeCase :: Bool
enterpriseBridgeCase = sendNotice "ALERT" "disk" ("kafka>"++) == "kafka>ALERT:disk" && sendNotice "REMINDER" "backup" ("queue>"++) == "queue>REMINDER:backup"

-- Enterprise Facade
enterpriseFacadeCase :: Bool
enterpriseFacadeCase =
  let crm :: Int -> String
      crm i = "crm:create:" ++ show i
      billing :: Int -> String
      billing i = "billing:open:" ++ show i
  in crm 77 ++ ">" ++ billing 77 == "crm:create:77>billing:open:77"

-- Broker
type Broker = [(String, String -> String)]
callBroker :: Broker -> String -> String -> String
callBroker b service arg = fromMaybe (error "service not found") (lookup service b) arg
brokerCase :: Bool
brokerCase = let b=[("inventory",\k->"inventory:"++k++"=7"),("customer",\k->"customer:"++k++"=active")] in callBroker b "inventory" "sku-1"=="inventory:sku-1=7" && callBroker b "customer" "17"=="customer:17=active"

-- Message Bus
data Message = Message String Int
type MessageHandler = Message -> String
sendBus :: [MessageHandler] -> Message -> [String]
sendBus hs m = map ($ m) hs
messageBusCase :: Bool
messageBusCase = let audit (Message t i)="audit:"++t++":"++show i; billing (Message t i)="billing:"++t++":"++show i in sendBus [audit,billing] (Message "order-created" 42)==["audit:order-created:42","billing:order-created:42"]

-- Service Locator
serviceLocatorCase :: Bool
serviceLocatorCase = let loc=[("email",("email>"++)),("audit",("audit>"++))] in fromMaybe id (lookup "email" loc) "a@example.test"=="email>a@example.test" && fromMaybe id (lookup "audit" loc) "created"=="audit>created"

-- Active Object
activeObjectCase :: Bool
activeObjectCase = let queue=[(+3),(*4)]; before=(0 :: Int); after=foldl (flip ($)) before queue in before==0 && after==12

-- Monitor Object: state and synchronization are encapsulated in an MVar.
monitorObjectCase :: IO Bool
monitorObjectCase = do
  m <- newMVar (0 :: Int)
  done1 <- newEmptyMVar
  done2 <- newEmptyMVar
  _ <- forkIO (modifyMVar_ m (pure . (+2)) >> putMVar done1 ())
  _ <- forkIO (modifyMVar_ m (pure . (+3)) >> putMVar done2 ())
  takeMVar done1 >> takeMVar done2
  (==5) <$> readMVar m

-- Half-Sync / Half-Async
halfSyncHalfAsyncCase :: Bool
halfSyncHalfAsyncCase = let queued=["job-1","job-2","job-3"]; processed=map ("done:"++) queued in processed==["done:job-1","done:job-2","done:job-3"]

-- Leader / Followers
leaderFollowersCase :: Bool
leaderFollowersCase = let workers=cycle ["worker-1","worker-2","worker-3"]; events=["event-a","event-b","event-c"]; handled=zipWith (\w e->w++":"++e) workers events; nextWorker=workers!!3 in handled==["worker-1:event-a","worker-2:event-b","worker-3:event-c"] && nextWorker=="worker-1"

-- Client-Server
data Request = Request String
data Response = Response Int String
serve :: Request -> Response
serve (Request "sku-1") = Response 200 "stock=7"
serve _ = Response 404 "missing"
clientServerCase :: Bool
clientServerCase = case serve (Request "sku-1") of Response s b -> s==200 && b=="stock=7"

-- Peer-to-Peer
data Peer = Peer String
peerSend :: Peer -> Peer -> String -> String
peerSend (Peer a) (Peer b) payload = a ++ ">" ++ b ++ ":" ++ payload
peerToPeerCase :: Bool
peerToPeerCase = let a=Peer "peer-a" in [peerSend a (Peer "peer-b") "block-42",peerSend a (Peer "peer-c") "block-42"]==["peer-a>peer-b:block-42","peer-a>peer-c:block-42"]

-- Publish-Subscribe
type Subscription = (String, Int -> String)
publishTopic :: [Subscription] -> String -> Int -> [String]
publishTopic subs topic value = [f value | (t,f) <- subs, t==topic]
publishSubscribeCase :: Bool
publishSubscribeCase = let s=[("order",("warehouse:"++) . show),("order",("analytics:"++) . show)] in publishTopic s "order" 51==["warehouse:51","analytics:51"]

-- Distributed Proxy
type StockService = String -> Int
stockProxy :: StockService -> StockService
stockProxy remote sku = remote sku
distributedProxyCase :: Bool
distributedProxyCase = stockProxy (const 7) "sku-1" == 7

-- Presentation-Abstraction-Control
data Agent = Agent String Int
agentView :: Agent -> String
agentView (Agent n v) = n ++ ":view=" ++ show v
pacCase :: Bool
pacCase = agentView (Agent "child" 42)=="child:view=42" && agentView (Agent "root" 42)=="root:view=42"

-- Model-View-Presenter
presentIncrement :: CounterModel -> (CounterModel,String)
presentIncrement m = let m'=controllerIncrement m in (m',viewCounter m')
mvpCase :: Bool
mvpCase = let (m,v)=presentIncrement (CounterModel 0) in counterValue m==1 && v=="count=1"

-- Document-View
data Document = Document String Int
editorView :: Document -> String
editorView (Document title words') = "editor:"++title++":"++show words'
summaryView :: Document -> String
summaryView (Document title _) = "summary:"++title
documentViewCase :: Bool
documentViewCase = let d=Document "Final" 120 in editorView d=="editor:Final:120" && summaryView d=="summary:Final"

-- Active Record
data PersonRecord = PersonRecord Int String deriving (Eq,Show)
saveRecord :: [PersonRecord] -> PersonRecord -> [PersonRecord]
saveRecord table record = record : filter (\(PersonRecord i _) -> i /= recordId record) table
  where recordId (PersonRecord i _) = i
loadRecord :: [PersonRecord] -> Int -> Maybe PersonRecord
loadRecord table wanted = case filter (\(PersonRecord i _) -> i==wanted) table of r:_ -> Just r; [] -> Nothing
activeRecordCase :: Bool
activeRecordCase = loadRecord (saveRecord [] (PersonRecord 7 "Ada")) 7 == Just (PersonRecord 7 "Ada")

-- Data Mapper
data Person = Person Int String deriving (Eq,Show)
data PersonRow = PersonRow String String
mapToRow :: Person -> PersonRow
mapToRow (Person i n)=PersonRow ("person:"++show i) n
mapFromRow :: PersonRow -> Person
mapFromRow (PersonRow _ n)=Person 8 n
dataMapperCase :: Bool
dataMapperCase = let row@(PersonRow key _)=mapToRow (Person 8 "Grace"); Person _ n=mapFromRow row in key=="person:8" && n=="Grace"

-- Unit of Work
commitUnit :: [Int] -> [(Int,Int)] -> [Int]
commitUnit values changes = [v + sum [d | (j,d)<-changes,j==i] | (i,v)<-zip [0..] values]
unitOfWorkCase :: Bool
unitOfWorkCase = let before=[10,20]; after=commitUnit before [(0,5),(1,-3)] in before==[10,20] && after==[15,17]

-- Repository
newtype PersonRepository = PersonRepository [Person]
findPerson :: PersonRepository -> Int -> Maybe Person
findPerson (PersonRepository ps) wanted = case filter (\(Person i _)->i==wanted) ps of p:_ -> Just p; [] -> Nothing
repositoryCase :: Bool
repositoryCase = findPerson (PersonRepository [Person 9 "Linus"]) 9 == Just (Person 9 "Linus")

-- Dependency Injection
type Sender = String -> String
greet :: Sender -> String -> String
greet sender name = sender name
dependencyInjectionCase :: Bool
dependencyInjectionCase = greet ("smtp:"++) "Ada"=="smtp:Ada" && greet ("fake:"++) "Ada"=="fake:Ada"

-- Lazy Initialization
data LazyValue a = Pending (() -> a) | Ready a
getLazy :: LazyValue a -> (a, LazyValue a)
getLazy (Ready a) = (a, Ready a)
getLazy (Pending f) = let a=f () in (a, Ready a)
lazyInitializationCase :: Bool
lazyInitializationCase = let (a,l1)=getLazy (Pending (const "resource-ready")); (b,l2)=getLazy l1 in a=="resource-ready" && b=="resource-ready" && case l2 of { Ready _ -> True; _ -> False }

-- Object Pool
data ObjectPool = ObjectPool [Int] Int
acquire :: ObjectPool -> (Int,ObjectPool)
acquire (ObjectPool (x:xs) n)=(x,ObjectPool xs n)
acquire (ObjectPool [] n)=let v=n+1 in (v,ObjectPool [] v)
release :: Int -> ObjectPool -> ObjectPool
release v (ObjectPool xs n)=ObjectPool (v:xs) n
objectPoolCase :: Bool
objectPoolCase = let (a,p1)=acquire (ObjectPool [] 0); (b,p2)=acquire p1; p3=release a p2; (c,_)=acquire p3 in (a,b,c)==(1,2,1)

-- Null Object
type Logger = String -> String
nullObjectCase :: Bool
nullObjectCase = let real msg="logged:"++msg; nullLogger _="" in real "processed:item-1"=="logged:processed:item-1" && nullLogger "processed:item-1"==""

pureCases :: [Bool]
pureCases = [ commandCase, interpreterCase, iteratorCase, mementoCase, observerCase, stateCase, strategyCase, templateMethodCase, visitorCase
            , mvcCase, mvvmCase, microkernelCase, microservicesCase, enterpriseAdapterCase, enterpriseBridgeCase, enterpriseFacadeCase, brokerCase, messageBusCase, serviceLocatorCase
            , activeObjectCase, halfSyncHalfAsyncCase, leaderFollowersCase, clientServerCase, peerToPeerCase, publishSubscribeCase, distributedProxyCase, pacCase, mvpCase, documentViewCase
            , activeRecordCase, dataMapperCase, unitOfWorkCase, repositoryCase, dependencyInjectionCase, lazyInitializationCase, objectPoolCase, nullObjectCase ]

main :: IO ()
main = do
  must (length pureCases == 37)
  forM_ pureCases must
  mediatorOk <- mediatorCase
  must mediatorOk
  monitorOk <- monitorObjectCase
  must monitorOk
  putStrLn "Haskell pattern sweep: 39/39 examples passed"
